extensions/interactions/AlgebraicExpressionInput/AlgebraicExpressionInput.py

```diff
@@ -53,3 +56,202 @@ class AlgebraicExpressionInput(base.BaseInteraction):
+        for answer_group in answer_groups:
+            base_answer_group_proto = (
+                answer_group.to_android_answer_group_proto())
+            rules_spec_proto = cls._convert_rule_specs_to_proto(
+                answer_group.rule_specs)
+            answer_group_list_proto.append(
+                state_pb2.AlgebraicExpressionInputInstanceDto.AnswerGroupDto(
+                    base_answer_group=base_answer_group_proto,
+                    rule_specs=rules_spec_proto
+                )
+            )
+
+        return answer_group_list_proto
+
+    @classmethod
+    def _convert_rule_specs_to_proto(cls, rule_specs_list):
+        """Creates a RuleSpecDto proto object.
+
+        Args:
+            rule_specs_list: list(RuleSpec). List of rule specifications.
+
+        Returns:
+            list(RuleSpecDto). The proto object list.
+        """
+        rule_specs_list_proto = []
+
+        rule_type_to_proto_func_mapping = {
+            'MatchesExactlyWith': (
+                cls._convert_matches_exactly_rule_spec_to_proto),
+            'IsEquivalentTo': cls._convert_is_equivalent_rule_spec_to_proto,
+            'MatchesUpToTrivialManipulations': (
+                cls._convert_matches_up_to_trivial_manipulation_rule_spec_to_proto     # pylint: disable=line-too-long
+            )
+        }
+
+        rule_type_to_proto_mapping = {
+            'MatchesExactlyWith': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    matches_exactly_with=x)),
+            'IsEquivalentTo': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    is_equivalent_to=x)),
+            'MatchesUpToTrivialManipulations': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    matches_upTo_trivial_manipulations=x))
+        }
+
+        # As AlgebraicExpressionInputInstanceDto does not contain all the
+        # rules, a check is necessary to create proto for only those rules
+        # which are required in proto creation.
+        for rule_spec in rule_specs_list:
+            rule_type = rule_spec.rule_type
+            if rule_type in rule_type_to_proto_func_mapping:
+                rule_proto = (
+                    rule_type_to_proto_func_mapping[rule_type](rule_spec.inputs)
+                )
+                rule_specs_list_proto.append(
+                    rule_type_to_proto_mapping[rule_type](rule_proto)
+                )
+
+        return rule_specs_list_proto
+
+    @classmethod
+    def _convert_matches_exactly_rule_spec_to_proto(cls, inputs):
+        """Creates a proto object for MatchesExactlyWithSpecDto.
+
+        Args:
+            inputs: dict. The input items.
+
+        Returns:
+            MatchesExactlyWithSpecDto. The proto object.
+        """
+        return RULE_SPEC_DTO.MatchesExactlyWithSpecDto(
+            algebraic_expression=inputs['x']
+        )
```

Comment:
I think these are a bit redundant, you can just put the `RULE_SPEC_DTO.MatchesExactlyWithSpecDto` into the `rule_type_to_proto_func_mapping` and get rid of this function.